### PR TITLE
dns: introduce ResolutionStatus for ResolveCb and fix #9927

### LINKED ISSUE #9927: DNS timeouts cause no healthy upstream
*Title*: DNS timeouts cause no healthy upstream

[optional *Relevant Links*:]
I have a test case outlined here on how to reproduce this. I have verified this happens on master branch too.

**Bug Template**

*Description*:
In the event that DNS resolver has transient errors (i.e timeout) envoy currently doesn't check the c-ares status of `SUCCESS` before overriding the address_list with an empty array. We have observed in production cases where we have healthy hosts, and the next time a DNS query happens if it were to timeout, our service ends up going down.

*Repro steps*:
Test case to reproduce. a simple iptables block on the DNS resolver will do.

https://github.com/chrisgoffinet/envoy-dns
Patch: https://github.com/chrisgoffinet/envoy/commit/950c734033da628b1ca7b7dc2656fdebea75128c

### PR DESCRIPTION
**Pull request by junr03**
Description: this PR introduces ResolutionStatus to the ResolveCb. This status can be used to determine if DnsResolution was successful or not, and lets the callback target adjust accordingly. It is used to fix #9927. Additionally, it can be used to fix
https://github.com/lyft/envoy-mobile/issues/673 in a subsequent PR.
Risk Level: med. This PR changes the behavior of Strict Dns Cluster. The introduced argument is not used anywhere, so all behavior stays the same.
Testing: Updated tests to pass the new argument to the callback. But behavior stays the same.
Docs Changes: updated config docs and discovery docs.
Release Notes: added a release note.

Fixes #9927 

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
@mattklein123 I finally got time to come back to this. I started by just introducing the new argument for the callback, but leaving all behavior unchanged. I didn't want to conflate this with the two independent changes (don't update hosts on failure, allow different timeouts on failure) needed for the linked issues.

Let me know if you'd rather look at it all together.

File: include/envoy/network/dns.h [hunk 1]
```diff
@@ -45,12 +45,17 @@ class DnsResolver {
...
   virtual ~DnsResolver() = default;
 
+  /**
+   * Final status for a DNS resolution.
+   */
+  enum class ResolutionStatus { Success, Failure };
+
   /**
    * Called when a resolution attempt is complete.
-   * @param response supplies the list of resolved IP addresses and TTLs. The list will be empty if
```
Comment (junr03) [inline include/envoy/network/dns.h]: > The list will be empty if the resolution failed.

Most of the callback targets actually use this implicit "failure" hint. The strict dns cluster does not, which is why #9927 happens.

I am perusing through c-ares and trying to assert if according to RFC we can reasonably expect ARES_SUCCESS with no resolved addresses. @mattklein123 do you know off of the top of your mind?

If that is the case, then we don't _necessarily_ need to add the status. Just make sure the callback targets behave as we expect when the address list is empty.

Comment (mattklein123) [reply include/envoy/network/dns.h]: Yes it is definitely possible for a DNS resolution to return nothing and be successful, so we definitely need to handle this differently. What we do today is really just a hack. If higher layer code *additionally* wants to stabilize hosts in the case of 0 legitimate responses that is fine (e.g. I think logical DNS should do this, but strict DNS should not).

File: source/common/network/dns_impl.cc [hunk 2]
```diff
@@ -89,7 +89,7 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
     // callback_ target _should_ still be around. In that case, raise the callback_ so the target
     // can be done with this query and initiate a new one.
     if (!cancelled_) {
-      callback_({});
+      callback_(ResolutionStatus::Failure, {});
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: Should you have another status like cancelled/destructing/etc.? This is not quite a failure? I'm not sure it matters to higher layer code but it might?

File: test/common/network/dns_impl_test.cc [hunk 3]
```diff
@@ -480,11 +480,12 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, DnsImplTest,
...
 // destruction.
 TEST_P(DnsImplTest, DestructPending) {
-  ActiveDnsQuery* query = resolver_->resolve("", DnsLookupFamily::V4Only,
-                                             [&](std::list<DnsResponse>&& results) -> void {
-                                               FAIL();
-                                               UNREFERENCED_PARAMETER(results);
-                                             });
+  ActiveDnsQuery* query = resolver_->resolve(
+      "", DnsLookupFamily::V4Only,
+      [&](DnsResolver::ResolutionStatus, std::list<DnsResponse>&& results) -> void {
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: Can you actually add expectations around resolution status in this file where appropriate?

File: test/common/upstream/logical_dns_cluster_test.cc [hunk 4]
```diff
@@ -156,10 +159,12 @@ class LogicalDnsClusterTest : public testing::Test {
     expectResolve(Network::DnsLookupFamily::V4Only, expected_address);
     resolve_timer_->invokeCallback();
 
-    // Empty should not cause any change.
+    // Failure should not cause any change.
+    // TODO(junr03): make sure to have coverage here. Why does strict change but logical not? Why
+    // this inconsistency? Updated in subsequent change.
```
Comment (mattklein123) [inline test/common/upstream/logical_dns_cluster_test.cc]: See my other comment. IMO logical DNS should "stabilize" but strict should not in the case of a successful nothing response, but let's make sure to clearly document this.

File: test/common/network/dns_impl_test.cc (hunk 3 above)
Comment (junr03) [reply test/common/network/dns_impl_test.cc]: Yep! I didn't want to spend too much time creating all of this until we discussed https://github.com/envoyproxy/envoy/pull/10137#discussion_r382861884. Will update

File: source/common/network/dns_impl.cc (hunk 2 above)
Comment (junr03) [reply source/common/network/dns_impl.cc]: I thought about this when I first introduced the destruction behavior. My first opinion is that higher layers only pivot behavior based on success/failure so I was going to keep it at that level of complexity for this change, and the subsequent PR where I fix #9927. 

Then I was going to explore the need to add a `Destruction` value that allow us to set tighter timers than on "regular failure" when fixing for https://github.com/lyft/envoy-mobile/issues/673. But I am not sure if we need different behavior based on what particular failure scenario the resolver experienced. By the same token, people might want even more granularity in the future?

In short: I did not want to commit to more granular status in this initial PR, and I don't think it is going to be that bad to add more granularity, if desired, in subsequent PRs.

File: include/envoy/network/dns.h (hunk 1 above)
Comment (junr03) [reply include/envoy/network/dns.h]: Yep, agreed it is a hack. But I wanted to assert if it was a hack based on actual possible behavior; c-ares would only have an empty list with non-ARES_SUCCESS status. It wasn't immediately clear to me from my reading. Given that is not the case, I'll go ahead and continue with this change. Thanks!

**Comment by junr03** [comment]
@mattklein123 updated. I have updated dns_impl_test for expectations around status in different conditions. I also updated higher level constructs to use status, but I _didn't_ change any behavior yet based on this new information (other than adding more relevant stats where possible). I plan on addressing https://github.com/envoyproxy/envoy/pull/10137#discussion_r383415588 and hence #9927 in a subsequent PR. Let me know if you'd rather see it here.

File: test/common/upstream/logical_dns_cluster_test.cc (hunk 4 above)
Comment (junr03) [reply test/common/upstream/logical_dns_cluster_test.cc]: chatted about this offline, as I was thinking that the suggested behavior change was a little odd. I am now clear on the expectation.

We also agreed that at this point we can just do #9927 here. I will update this today.

File: source/common/upstream/logical_dns_cluster.cc [hunk 5]
```diff
@@ -100,23 +100,24 @@ void LogicalDnsCluster::startResolve() {
...
+        // not stabilize back to 0 hosts.
+        if (status == Network::DnsResolver::ResolutionStatus::Success && !response.empty()) {
+          info_->stats().update_success_.inc();
           // TODO(mattklein123): Move port handling into the DNS interface.
           ASSERT(response.front().address_ != nullptr);
           Network::Address::InstanceConstSharedPtr new_address =
               Network::Utility::getAddressWithPort(*(response.front().address_),
                                                    Network::Utility::portFromTcpUrl(dns_url_));
 
-          if (respect_dns_ttl_ && response.front().ttl_ != std::chrono::seconds(0)) {
```
Comment (junr03) [inline source/common/upstream/logical_dns_cluster.cc]: moved this for consistency with the strict dns cluster code. But happy to move back up.

File: test/common/upstream/logical_dns_cluster_test.cc [hunk 6]
```diff
@@ -272,6 +288,88 @@ TEST_P(LogicalDnsParamTest, ImmediateResolve) {
...
+
+  // Therefore, a subsequent failure should get a [0,base * 1] refresh.
+  ON_CALL(random_, random()).WillByDefault(Return(8000));
+  EXPECT_CALL(*resolve_timer_, enableTimer(std::chrono::milliseconds(1000), _));
+  dns_callback_(Network::DnsResolver::ResolutionStatus::Failure, {});
+
+  tls_.shutdownThread();
+}
+
+TEST_F(LogicalDnsParamTest, TtlAsDnsRefreshRate) {
```
Comment (junr03) [inline test/common/upstream/logical_dns_cluster_test.cc]: Adding coverage that did not exist.

File: test/common/network/dns_impl_test.cc [hunk 7]
```diff
@@ -706,7 +742,10 @@ TEST_P(DnsImplTest, RemoteAsyncLookup) {
   server_->addHosts("some.good.domain", {"201.134.56.7"}, RecordType::A);
   std::list<Address::InstanceConstSharedPtr> address_list;
   EXPECT_NE(nullptr, resolver_->resolve("some.bad.domain", DnsLookupFamily::Auto,
-                                        [&](std::list<DnsResponse>&& results) -> void {
+                                        [&](DnsResolver::ResolutionStatus status,
+                                            std::list<DnsResponse>&& results) -> void {
+                                          // Failure f
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: ?

File: test/common/network/dns_impl_test.cc [hunk 8]
```diff
@@ -715,7 +754,8 @@ TEST_P(DnsImplTest, RemoteAsyncLookup) {
   EXPECT_TRUE(address_list.empty());
 
   EXPECT_NE(nullptr, resolver_->resolve("some.good.domain", DnsLookupFamily::Auto,
-                                        [&](std::list<DnsResponse>&& results) -> void {
+                                        [&](DnsResolver::ResolutionStatus,
+                                            std::list<DnsResponse>&& results) -> void {
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: I'm not sure what the best way would be but is there some way to simplify all of thee calls to make sure have a success/failure expectation on all of them? Perhaps some helper function/object that does the resolve, does the expectations, and stores the resolution list?

File: test/common/network/dns_impl_test.cc [hunk 9]
```diff
@@ -635,279 +649,174 @@ TEST_P(DnsImplTest, DestroyChannelOnRefused) {
...
-
+  EXPECT_NE(nullptr, resolveWithExpectations("some.good.domain", DnsLookupFamily::Auto,
+                                             DnsResolver::ResolutionStatus::Success,
+                                             {"201.134.56.7"}, {}, absl::nullopt));
   dispatcher_->run(Event::Dispatcher::RunType::Block);
   EXPECT_FALSE(peer_->isChannelDirty());
-  EXPECT_TRUE(hasAddress(address_list, "201.134.56.7"));
-}
-
-TEST_P(DnsImplTest, DnsIpAddressVersion) {
```
Comment (junr03) [inline test/common/network/dns_impl_test.cc]: moved above

File: test/common/network/dns_impl_test.cc (hunk 8 above)
Comment (junr03) [reply test/common/network/dns_impl_test.cc]: folded everything into `resolveWithExpectations`

File: test/common/network/dns_impl_test.cc [hunk 10]
```diff
@@ -495,117 +523,107 @@ TEST_P(DnsImplTest, DestructPending) {
...
 // Validate exception behavior during c-ares callbacks.
 TEST_P(DnsImplTest, CallbackException) {
   // Force immediate resolution, which will trigger a c-ares exception unsafe
   // state providing regression coverage for #4307.
-  EXPECT_EQ(nullptr, resolver_->resolve(
-                         "1.2.3.4", DnsLookupFamily::V4Only,
-                         [&](const std::list<DnsResponse> &
-                             /*results*/) -> void { throw EnvoyException("Envoy exception"); }));
+  EXPECT_EQ(nullptr,
+            resolver_->resolve(
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: qq: Can't we use the new resolve with expectations everywhere?

Comment (junr03) [reply test/common/network/dns_impl_test.cc]: yes! I thought it might be overkill given that there were far fewer call sites. But if I'm cleaning up something why not go the entire way :). Now everything (for real) is folded into helper functions.

**Comment by junr03** [comment]
Looks like the circle machines have a funny hosts file?

```
[----------] 1 test from IpVersions/DnsImplTest
[ RUN      ] IpVersions/DnsImplTest.LocalLookup/IPv4
test/common/network/dns_impl_test.cc:470: Failure
Expected equality of these values:
  expected_results
    Which is: { "127.0.0.1" }
  address_as_string_list
    Which is: { "127.0.0.1", "127.0.0.1" }
```

**Comment by mattklein123** [comment]
Thanks for the fix. If that's the CI error you are seeing it seems like it must have to do with your change?

**Comment by junr03** [comment]
Yep, definitely has to do with my change. I just wanted to make sure that I was surfacing something latent in circle ci and not an unintended regression. Previously the `LocalLookup` test only asserted that `127.0.0.1` was contained in the resolved address list, not exact match like now. I opened https://github.com/envoyproxy/envoy/pull/10197 to verify. It looks like our circle CI setup ends up with a hosts file with a double entry for localhost. https://app.circleci.com/jobs/github/envoyproxy/envoy/324116.
